src/PooledArrays.jl

```diff
@@ -439,40 +446,49 @@ Base.convert(::Type{Array}, pa::PooledArray{T, R, N}) where {T, R, N} = convert(
 ##
 ##############################################################################
 
-# Scalar case
-Base.@propagate_inbounds function Base.getindex(pav::PooledArrOrSub, I::Integer...)
-    idx = DataAPI.refarray(pav)[I...]
-    iszero(idx) && throw(UndefRefError())
-    return @inbounds DataAPI.refpool(pav)[idx]
+# We need separate functions due to dispatch ambiguities
+
+for T in (PooledArray, SubArray{<:Any, <:Any, <:PooledArray})
+    @eval Base.@propagate_inbounds function Base.getindex(A::$T, I::Integer...)
+        idx = DataAPI.refarray(A)[I...]
+        iszero(idx) && throw(UndefRefError())
+        return @inbounds DataAPI.refpool(A)[idx]
+    end
+
+    @eval Base.@propagate_inbounds function Base.getindex(A::$T, I::Union{Real, AbstractVector}...)
+        # make sure we do not increase A.refcount in case creation of newrefs fails
+        newrefs = getindex(DataAPI.refarray(A), I...)
```

Comment:
For consistency with previous method:
```suggestion
        newrefs = DataAPI.refarray(A)[I...]
```